File: src/main/java/de/pcrp/bot/common/WordFilter.java
package de.pcrp.bot.common;

/**
 * Wortfilter für vulgäre Kraftausdrücke (Deutsch + Englisch).
 * Nachrichten mit Treffern werden sofort gelöscht.
 * Leetspeak-Normalisierung ist integriert (z.B. "f1ck" → "fick").
 */
public final class WordFilter {

    private static final String[] BANNED_WORDS = {
        // Deutsch
        "hurensohn", "hurentochter", "hure", "nutte", "fotze", "wichser",
        "wixer", "wixxer", "schlampe", "missgeburt", "bastard", "arschloch",
        "drecksau", "dreckskerl", "spast", "spasti", "mongo", "behindert",
        "fick dich", "fickt euch", "verfickt", "ficker", "kanake", "untermensch",
        // Englisch
        "fuck", "fucking", "fucker", "motherfucker", "bitch", "asshole",
        "cunt", "slut", "whore", "dickhead", "retard", "faggot", "nigger",
        "nigga", "cocksucker", "son of a bitch",
    };

    private WordFilter() {}

    /** Prüft, ob der Text einen vulgären Kraftausdruck enthält. */
    public static boolean containsBannedWord(String content) {
        return getMatchedWord(content) != null;
    }

    /**
     * Gibt das erste gefundene verbotene Wort zurück, oder null wenn keines gefunden wurde.
     * Wird vom Log-System genutzt, um den gefundenen Begriff zu protokollieren.
     */
    public static String getMatchedWord(String content) {
        if (content == null || content.isBlank()) return null;
        String normalized = normalize(content);
        for (String word : BANNED_WORDS) {
            if (normalized.contains(word)) return word;
        }
        return null;
    }

    /** Kleinbuchstaben + einfache Leetspeak-Umschreibungen auflösen. */
    private static String normalize(String input) {
        StringBuilder sb = new StringBuilder(input.length());
        for (char c : input.toLowerCase().toCharArray()) {
            sb.append(switch (c) {
                case '0' -> 'o';
                case '1' -> 'i';
                case '3' -> 'e';
                case '4' -> 'a';
                case '5' -> 's';
                case '7' -> 't';
                case '@' -> 'a';
                case '$' -> 's';
                default  -> c;
            });
        }
        return sb.toString();
    }
}

File: pcrp-bot/src/PCRP.Bot/Common/WordFilter.cs
using System.Text;

namespace PCRP.Bot.Common;

/// <summary>
/// Wortfilter für vulgäre Kraftausdrücke (Deutsch + Englisch).
/// Nachrichten mit Treffern werden sofort gelöscht.
/// </summary>
public static class WordFilter
{
    private static readonly string[] BannedWords =
    {
        // Deutsch
        "hurensohn", "hurentochter", "hure", "nutte", "fotze", "wichser",
        "wixer", "wixxer", "schlampe", "missgeburt", "bastard", "arschloch",
        "drecksau", "dreckskerl", "spast", "spasti", "mongo", "behindert",
        "fick dich", "fickt euch", "verfickt", "ficker", "kanake", "untermensch",
        // Englisch
        "fuck", "fucking", "fucker", "motherfucker", "bitch", "asshole",
        "cunt", "slut", "whore", "dickhead", "retard", "faggot", "nigger",
        "nigga", "cocksucker", "son of a bitch",
    };

    /// <summary>Prüft, ob der Text einen vulgären Kraftausdruck enthält.</summary>
    public static bool ContainsBannedWord(string content)
    {
        if (string.IsNullOrWhiteSpace(content))
            return false;

        var normalized = Normalize(content);
        return BannedWords.Any(word => normalized.Contains(word, StringComparison.Ordinal));
    }

    /// <summary>Kleinbuchstaben + einfache Leetspeak-Umschreibungen auflösen.</summary>
    private static string Normalize(string input)
    {
        var sb = new StringBuilder(input.Length);
        foreach (var c in input.ToLowerInvariant())
        {
            sb.Append(c switch
            {
                '0' => 'o',
                '1' => 'i',
                '3' => 'e',
                '4' => 'a',
                '5' => 's',
                '7' => 't',
                '@' => 'a',
                '$' => 's',
                _ => c,
            });
        }
        return sb.ToString();
    }
}
